### ISSUE TITLE
[bug report] ConvTranspose is not padding output with zero

### ISSUE THREAD

**Issue by cosimo-dw** [pytorch/pytorch#8816]
## Issue description

From the documentation of [ConvTranspose2d](https://pytorch.org/docs/stable/nn.html#torch.nn.ConvTranspose2d), `output_padding` controls the number of rows and columns being zero-padded to adjust the output shape. However, the values of the last rows/columns are not zero. They are neither padded using reflection padding nor repeated padding.

## Code example

```
>>> import torch
>>> conv = torch.nn.ConvTranspose2d(1,1,3,stride=2,padding=1,output_padding=1)
>>> dummy = torch.ones(1,1,1,1)
>>> out = conv(dummy)
>>> out

(0 ,0 ,.,.) = 
 -0.4996 -0.5142
 -0.2036 -0.4360
[torch.FloatTensor of size (1,1,2,2)]
```

The output should be
```
(0 ,0 ,.,.) = 
 -0.4996 0
       0 0
[torch.FloatTensor of size (1,1,2,2)]
```

## System Info
PyTorch version: 0.4.0a0+5bbeb55
Is debug build: No
CUDA used to build PyTorch: 9.1

OS: Mac OSX 10.13.5
GCC version: Could not collect
CMake version: version 3.9.4

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy (1.14.5)
[conda] torch                     0.4.0a0+5bbeb55           <pip>

**Comment by soumith** [comment]
I think you misunderstand what `output_padding` does, and the documentation entry for the output_padding section should be improved / corrected.
There is a note in the ConvTranspose docs section that elaborates what `output_padding` actually does. It's padding that is effectively done at the right and/or bottom of the image. See the explanation below:

<img width="693" alt="screen shot 2018-06-22 at 8 55 47 pm" src="https://user-images.githubusercontent.com/1310570/41804098-b891a0d8-765e-11e8-929d-0e4dc773dba6.png">

**Comment by cosimo-dw** [comment]
I see. So `output_padding` actually adds more zeros to the input, to ensure that the output shape matches its description.

Thanks for clarifying that and it would be better to have the documentation improved.

**Comment by SsnL** [comment]
@cosimo-dw It doesn't adds zeros to input either. It is used to resolve the ambiguity when one have `stride > 1`. E.g., consider a 3x3 conv with stride = 2. You would have:
1. `3x3 input => 1x1 output`
2. `4x4 input => 1x1 output`
3. `5x5 input => 2x2 output`
4. `6x6 input => 2x2 output`.

So it is a many-to-one mapping. For convT to "revert" this shape change, it needs an extra argument like `output_padding` to know (e.g. from `2x2 input` whether it should get `5x5 output` or `6x6 output`).

**Comment by cosimo-dw** [comment]
@SsnL Sorry I am still a bit confused. If it is not padding input how does it control the size of the output? Can you point me out the source code of the implementation so that I can read it myself?

**Comment by soumith** [comment]
@cosimo-dw it is padding the input, on the right hand side and on the bottom.

See: https://github.com/pytorch/pytorch/blob/master/aten/src/THNN/generic/SpatialFullDilatedConvolution.c#L242-L243

Follow the adjH and adjW variables

**Comment by cosimo-dw** [comment]
@soumith Thank you.
@SsnL I was talking about the actual behavior to solve the ambiguity. For your example of `2x2 input`, without `output_padding` the output is `5x5`, while with `output_padding=1`, it should sample one more row (and column) of patches in the input image by extending the size of input by 1, so that the convoluted output size becomes `6x6`. It is controlled by the params of im2col in the code referred above.
